# Snakefile
# Tells which files to generate as specified by the rule 'all'

# Snakefile to quality check and estimate counts from RNA-Seq data on a reference genome (e.g. human genome hg19)
# Tells which files to generate as specified by the rule 'all'

import os

#########################
## Pipeline configuration
#########################

# configuration file
configfile: "config.yaml"

# Number of CPU (threads)
THREADS = config["threads"]

# Trimmomatic
ADAPTERFILE = config["trimmomatic"]["adapters"]
TRIMMOMATIC = config["trimmomatic"]["jarfile"]

##################
## Desired outputs
##################
FASTQC_REPORTS = expand("fastqc/trimmed/{sample}/{pair}/{sample}_{pair}_fastqc.html",sample=config["samples"],pair=["forward","reverse"])
STAR2PASS = "./star2pass/chrLength.txt"
GENECOUNTS = "results/gene_counts.txt"

rule all:
	input:
		STAR2PASS,
		GENECOUNTS
	message: "RNA-Seq pipeline has been successfully run"

########################
## version documentation
########################






################################################
## Count read alignments per feature (e.g. gene)
################################################
rule feature_counts:
    input:
        bams = expand("/zfs/scratch/mgalland_temp/neuro/mapped/{sample}_Aligned.sortedByCoord.out.bam",sample=config["samples"].keys()),
        gtf = config["annotations"]["gtf4genes"],
        genome = config["refseqs"]["genomefasta"]
    output:
        "results/gene_counts.txt"
    message:"summarizing read counts"
    shell:
        "featureCounts "
        "-t exon "		# specify how to count reads at feature level (e.g. exon)
        "-g gene_id " 		# specify how to count reads at meta-feature level (e.g. gene)
        "-T {THREADS} "
        "-G {input.genome} "	# ref genome sequence
        "-a {input.gtf} "	# annotation 
        "-o {output} "
        "{input}"               # can process more than one BAM file at once 
        

#############################
# Genome alignment using STAR
#############################
rule map_to_genome_using_STAR:
    input:
        ref = "./star2pass/",
        forward = "/zfs/scratch/mgalland_temp/neuro/trimmed/{sample}_forward.fastq",
        reverse = "/zfs/scratch/mgalland_temp/neuro/trimmed/{sample}_reverse.fastq"
    output:
        temp("/zfs/scratch/mgalland_temp/neuro/mapped/{sample}_Aligned.sortedByCoord.out.bam"),
        temp("/zfs/scratch/mgalland_temp/neuro/mapped/{sample}_Aligned.sortedByCoord.out.bam.bai"),
        "/zfs/scratch/mgalland_temp/neuro/mapped/{sample}_Log.final.out"
    message:"mapping the {wildcards.sample} reads to genome"
    params:
        prefix = "/zfs/scratch/mgalland_temp/neuro/mapped/{sample}_",
        maxmismatches = config["star"]["mismatches"],
        unmapped = config["star"]["unmapped"]	,
        multimappers = config["star"]["multimappers"],
        matchNminoverLread = config["star"]["matchminoverlengthread"] 	
    shell:
            "STAR --genomeDir {input.ref} "
            "--readFilesIn {input.forward} {input.reverse} "
            "--outFilterMultimapNmax {params.multimappers} "
            "--outFilterMismatchNmax {params.maxmismatches} "
            "--outFilterMatchNminOverLread {params.matchNminoverLread} "
            "--alignEndsType EndToEnd "
            "--runThreadN {THREADS} "
            "--outReadsUnmapped {params.unmapped} "
            "--outFileNamePrefix {params.prefix} "
            "--outSAMtype BAM SortedByCoordinate;"
            "samtools index {output[0]}"

#####################################################################
## STAR 2-pass: genome indexing + splice junctions database generation 
#####################################################################
rule star2pass_index:
    input:
        sjdb = "./star1pass/SJ.concatenated.out.tab",
        directory = "./star2pass/", 
        ref= config["refseqs"]["genomefasta"],
        gtf = config["annotations"]["gtf4genes"]
    output:
        STAR_2PASS = ["./star2pass/"+f for f in ["chrLength.txt","chrNameLength.txt","chrName.txt","chrStart.txt","Genome","genomeParameters.txt","SA","SAindex"]]
    message: "STAR 2nd pass: generating genome index"	
    shell:
        "STAR --runMode genomeGenerate "
        "--genomeDir {input.directory} "
        "--genomeFastaFiles {input.ref} "
        "--runThreadN {THREADS} "
        "--sjdbFileChrStartEnd {input.sjdb} "
        "--sjdbOverhang 99 "
        "--sjdbGTFfile {input.gtf};"
        "touch -h {output}"

rule create_star2pass_directory:
    output:"./star2pass/"
    message:"create directory for star2pass"
    shell:"mkdir -p ./star2pass/"

rule concatenate_sjdb:
    input:
        expand("./star1pass/{sample}_SJ.out.tab",sample=config["samples"].keys()),
    output:
        "./star1pass/SJ.concatenated.out.tab"
    message:"concatenating splice junctions from different samples "
    shell:"cat {input} >> {output}"

rule star1pass_align:
    input:
        forward = "/zfs/scratch/mgalland_temp/neuro/trimmed/{sample}_forward.fastq",
        reverse = "/zfs/scratch/mgalland_temp/neuro/trimmed/{sample}_reverse.fastq", 
        ref = "star_index/"
    output:
        "./star1pass/{sample}_SJ.out.tab",
        temp("./star1pass/{sample}_Aligned.out.sam")
    message:"STAR 1st pass: aligning {wildcards.sample} reads to generate splice junction files"
    params:"./star1pass/{sample}_"	
    shell: 		
        "STAR --runMode alignReads "
        "--genomeDir {input.ref} "
        "--readFilesIn {input.forward} {input.reverse} "
        "--outFileNamePrefix {params} "
        "--outFilterIntronMotifs RemoveNoncanonical "
        "--runThreadN {THREADS}"

 # sdjbOverhang specifies the length of the genomic sequence around the annotated junction to be used in constructing the splie junctions database. Ideally this length should be equal to ReadLength-1
rule star_genome_index:
    input:
        genome = config["refseqs"]["genomefasta"],
        gtf = config["annotations"]["gtf4genes"]
    output:
        "star_index/"
    message:"generation STAR genome index" 
    params:"star_index/"
    shell:
        "STAR --runMode genomeGenerate "
        "--genomeDir {params} "
        "--genomeFastaFiles {input.genome} "
        "--runThreadN {THREADS} "
        "--sjdbOverhang 99 "
        "-sjdbGTFfile {input.gtf}"

################################
## Fastqc reports after trimming
################################
rule fastqc_after_trimming:
    input:
        "/zfs/scratch/mgalland_temp/neuro/trimmed/{sample}_{pair}.fastq"
    output:
        "fastqc/trimmed/{sample}/{pair}/{sample}_{pair}_fastqc.html"
    message:"generating fastqc report for trimmed reverse {wildcards.sample} reads"
    params:
        "fastqc/trimmed/{sample}/{pair}/"
    shell:"fastqc --threads {THREADS} --outdir={params} {input}"
    

###########
## Trimming
###########
rule trimmomaticPaired:
    input:
        ADAPTERFILE,
        forward = lambda wildcards: config["fastqdir"] + config["samples"][wildcards.sample]["forward"],
        reverse = lambda wildcards: config["fastqdir"] + config["samples"][wildcards.sample]["reverse"]
    output:
        forward = "/zfs/scratch/mgalland_temp/neuro/trimmed/{sample}_forward.fastq",
        reverse = "/zfs/scratch/mgalland_temp/neuro/trimmed/{sample}_reverse.fastq",
        forwardUnpaired  = "/zfs/scratch/mgalland_temp/neuro/trimmed/{sample}_forward_unpaired.fastq",
        reverseUnpaired = "/zfs/scratch/mgalland_temp/neuro/trimmed/{sample}_reverse_unpaired.fastq",
    message:"Trimming {wildcards.sample} using Trimmomatic"
    log:"results/trimlogs/{wildcards.sample}_{wildcards.pair}.trimlog"
    params :
        seedMisMatches =            str(config['trimmomatic']['seedMisMatches']),
        palindromeClipTreshold =    str(config['trimmomatic']['palindromeClipTreshold']),
        simpleClipThreshhold =      str(config['trimmomatic']['simpleClipThreshold']),
        LeadMinTrimQual =           str(config['trimmomatic']['LeadMinTrimQual']),
        TrailMinTrimQual =          str(config['trimmomatic']['TrailMinTrimQual']),
        windowSize =                str(config['trimmomatic']['windowSize']),
        avgMinQual =                str(config['trimmomatic']['avgMinQual']),
        minReadLen =                str(config['trimmomatic']['minReadLength']),
        phred = 		    str(config["trimmomatic"]["phred"])
    shell:
        "java -jar {TRIMMOMATIC} PE {params.phred} -threads {THREADS} "
        "{input.forward} {input.reverse} "
        "{output.forward} {output.forwardUnpaired} "
        "{output.reverse} {output.reverseUnpaired} "
        "ILLUMINACLIP:{ADAPTERFILE}:{params.seedMisMatches}:{params.palindromeClipTreshold}:{params.simpleClipThreshhold} "
        "LEADING:{params.LeadMinTrimQual} "
        "TRAILING:{params.TrailMinTrimQual} "
        "SLIDINGWINDOW:{params.windowSize}:{params.avgMinQual} "
        "MINLEN:{params.minReadLen}"

